File: src/compiler/checker.format.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { createChecker } from "./checker.ts";
import { Diagnostics } from "./diagnostics.ts";
import { loadJdkStub } from "./jdkStub.ts";
import { createProgram } from "./program.ts";
import type { Uri } from "./program.ts";

const TOO_FEW = Diagnostics.Format_not_enough_arguments_0_1.code;
const TOO_MANY = Diagnostics.Format_too_many_arguments_0_1.code;
const BAD_TYPE = Diagnostics.Format_conversion_incompatible_0_1.code;

// Wrap a snippet in a method body so the format call resolves against the stub.
function diagnose(body: string): number[] {
  const program = createProgram();
  loadJdkStub(program);
  const uri = "file:///T.java" as Uri;
  const src = `import java.io.PrintWriter; import java.util.Locale; class C { void m(PrintWriter pw) { ${body} } }`;
  program.setOpenDocument(uri, src, 1);
  const checker = createChecker(program);
  const format = new Set<number>([TOO_FEW, TOO_MANY, BAD_TYPE]);
  return checker
    .getSemanticDiagnostics(program.getSourceFile(uri)!)
    .map(d => d.code)
    .filter(c => format.has(c));
}

test("String.format with too few arguments is flagged", () => {
  expect(diagnose('String.format("%s %s", "a");')).toContain(TOO_FEW);
});

test("String.format with too many arguments is flagged", () => {
  expect(diagnose('String.format("%s", "a", "b");')).toContain(TOO_MANY);
});

test("String.format with the right count is silent", () => {
  expect(diagnose('String.format("%s %s", "a", "b");')).toEqual([]);
});

test("%% and %n do not consume arguments", () => {
  expect(diagnose('String.format("%s 100%%%n", "a");')).toEqual([]);
});

test("positional index counts the highest reference", () => {
  expect(diagnose('String.format("%2$s %1$s", "a");')).toContain(TOO_FEW);
  expect(diagnose('String.format("%2$s %1$s", "a", "b");')).toEqual([]);
});

test("%d with a String argument is a type mismatch", () => {
  expect(diagnose('String.format("%d", "a");')).toContain(BAD_TYPE);
});

test("%d with an int argument is silent", () => {
  expect(diagnose('String.format("%d", 1);')).toEqual([]);
});

test("%s accepts any type", () => {
  expect(diagnose('String.format("%s", 1);')).toEqual([]);
});

test("System.out.printf is checked like format", () => {
  expect(diagnose('System.out.printf("%s %s", "a");')).toContain(TOO_FEW);
});

test("PrintWriter.format is checked", () => {
  expect(diagnose('pw.format("%d", "a");')).toContain(BAD_TYPE);
});

test("String.formatted uses the receiver as the format string", () => {
  expect(diagnose('"%s %s".formatted("a");')).toContain(TOO_FEW);
  expect(diagnose('"%s".formatted("a", "b");')).toContain(TOO_MANY);
});

test("Locale-first String.format overload is handled", () => {
  expect(diagnose('String.format(Locale.US, "%s %s", "a");')).toContain(TOO_FEW);
  expect(diagnose('String.format(Locale.US, "%s", "a");')).toEqual([]);
});

test("a non-literal format string is not analyzed", () => {
  expect(diagnose('String fmt = "%s %s"; String.format(fmt, "a");')).toEqual([]);
});

test("a malformed format string is not analyzed", () => {
  expect(diagnose('String.format("%z", "a");')).toEqual([]);
});

File: togo/internal/compiler/checker_format_test.go
// Port of src/compiler/checker.format.test.ts
package compiler

import "testing"

var (
	tooFew  = int(Diagnostics.FormatNotEnoughArguments01.Code)
	tooMany = int(Diagnostics.FormatTooManyArguments01.Code)
	badType = int(Diagnostics.FormatConversionIncompatible01.Code)
)

// formatDiagnose returns only the format-related diagnostic codes for a snippet
// wrapped in a method body (so the call resolves against the stub).
func formatDiagnose(body string) []int {
	program := NewProgram()
	LoadJdkStub(program)
	src := "import java.io.PrintWriter; import java.util.Locale; class C { void m(PrintWriter pw) { " + body + " } }"
	program.SetOpenDocument("file:///T.java", src, 1)
	checker := NewChecker(program)
	var out []int
	for _, d := range checker.GetSemanticDiagnostics(program.GetSourceFile("file:///T.java")) {
		code := int(d.Code)
		if code == tooFew || code == tooMany || code == badType {
			out = append(out, code)
		}
	}
	return out
}

func TestFormatTooFew(t *testing.T) {
	if !containsCode(formatDiagnose(`String.format("%s %s", "a");`), tooFew) {
		t.Error("want too-few warning")
	}
}

func TestFormatTooMany(t *testing.T) {
	if !containsCode(formatDiagnose(`String.format("%s", "a", "b");`), tooMany) {
		t.Error("want too-many warning")
	}
}

func TestFormatRightCountSilent(t *testing.T) {
	if got := formatDiagnose(`String.format("%s %s", "a", "b");`); len(got) != 0 {
		t.Errorf("want no warnings, got %v", got)
	}
}

func TestFormatPercentNewlineSilent(t *testing.T) {
	if got := formatDiagnose(`String.format("%s 100%%%n", "a");`); len(got) != 0 {
		t.Errorf("want no warnings, got %v", got)
	}
}

func TestFormatPositional(t *testing.T) {
	if !containsCode(formatDiagnose(`String.format("%2$s %1$s", "a");`), tooFew) {
		t.Error("want too-few warning")
	}
	if got := formatDiagnose(`String.format("%2$s %1$s", "a", "b");`); len(got) != 0 {
		t.Errorf("want no warnings, got %v", got)
	}
}

func TestFormatTypeMismatch(t *testing.T) {
	if !containsCode(formatDiagnose(`String.format("%d", "a");`), badType) {
		t.Error("want type-mismatch warning")
	}
	if got := formatDiagnose(`String.format("%d", 1);`); len(got) != 0 {
		t.Errorf("want no warnings, got %v", got)
	}
	if got := formatDiagnose(`String.format("%s", 1);`); len(got) != 0 {
		t.Errorf("want no warnings for %%s, got %v", got)
	}
}

func TestFormatPrintf(t *testing.T) {
	if !containsCode(formatDiagnose(`System.out.printf("%s %s", "a");`), tooFew) {
		t.Error("want too-few warning for printf")
	}
}

func TestFormatPrintWriter(t *testing.T) {
	if !containsCode(formatDiagnose(`pw.format("%d", "a");`), badType) {
		t.Error("want type-mismatch warning for PrintWriter.format")
	}
}

func TestFormatFormatted(t *testing.T) {
	if !containsCode(formatDiagnose(`"%s %s".formatted("a");`), tooFew) {
		t.Error("want too-few warning for formatted")
	}
	if !containsCode(formatDiagnose(`"%s".formatted("a", "b");`), tooMany) {
		t.Error("want too-many warning for formatted")
	}
}

func TestFormatLocaleOverload(t *testing.T) {
	if !containsCode(formatDiagnose(`String.format(Locale.US, "%s %s", "a");`), tooFew) {
		t.Error("want too-few warning for Locale overload")
	}
	if got := formatDiagnose(`String.format(Locale.US, "%s", "a");`); len(got) != 0 {
		t.Errorf("want no warnings, got %v", got)
	}
}

func TestFormatNonLiteralSilent(t *testing.T) {
	if got := formatDiagnose(`String fmt = "%s %s"; String.format(fmt, "a");`); len(got) != 0 {
		t.Errorf("want no warnings for non-literal, got %v", got)
	}
}

func TestFormatMalformedSilent(t *testing.T) {
	if got := formatDiagnose(`String.format("%z", "a");`); len(got) != 0 {
		t.Errorf("want no warnings for malformed, got %v", got)
	}
}
